server/src/main/java/org/opensearch/gateway/TransportNodesListGatewayStartedShardsBatch.java

```diff
@@ -0,0 +1,458 @@
+                            }
+                            Store.tryOpenIndex(shardPath.resolveIndex(), shardId, nodeEnv::shardLock, logger);
+                        } catch (Exception exception) {
+                            final ShardPath finalShardPath = shardPath;
+                            logger.trace(
+                                () -> new ParameterizedMessage(
+                                    "{} can't open index for shard [{}] in path [{}]",
+                                    shardId,
+                                    shardStateMetadata,
+                                    (finalShardPath != null) ? finalShardPath.resolveIndex() : ""
+                                ),
+                                exception
+                            );
+                            String allocationId = shardStateMetadata.allocationId != null ? shardStateMetadata.allocationId.getId() : null;
+                            shardsOnNode.put(
+                                shardId,
+                                new NodeGatewayStartedShards(allocationId, shardStateMetadata.primary, null, exception)
+                            );
+                            continue;
+                        }
+                    }
+
+                    logger.debug("{} shard state info found: [{}]", shardId, shardStateMetadata);
+                    String allocationId = shardStateMetadata.allocationId != null ? shardStateMetadata.allocationId.getId() : null;
+                    final IndexShard shard = indicesService.getShardOrNull(shardId);
+                    shardsOnNode.put(
+                        shardId,
+                        new NodeGatewayStartedShards(
+                            allocationId,
+                            shardStateMetadata.primary,
+                            shard != null ? shard.getLatestReplicationCheckpoint() : null
+                        )
+                    );
+                    continue;
+                }
+                logger.trace("{} no local shard info found", shardId);
+                shardsOnNode.put(shardId, new NodeGatewayStartedShards(null, false, null));
+            } catch (Exception e) {
+                shardsOnNode.put(
+                    shardId,
+                    new NodeGatewayStartedShards(null, false, null, new OpenSearchException("failed to load started shards", e))
+                );
+            }
+        }
+        return new NodeGatewayStartedShardsBatch(clusterService.localNode(), shardsOnNode);
+    }
+
+    /**
+     * This is used in constructing the request for making the transport request to set of other node.
+     * Refer {@link TransportNodesAction} class start method.
+     *
+     * @opensearch.internal
+     */
+    public static class Request extends BaseNodesRequest<Request> {
+        private final Map<ShardId, String> shardIdsWithCustomDataPath;
```

Comment:
create separate class for holding ShardId & customDataPath. And then store list of it here. This will help in future if  more attributes have to be added. Those can't be accommodated with current map.

Reply:
Created a class ShardAttributes in latest commit, and a List of ShardAttributes is used here in latest commit.